tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/AcquireTokenSilentTests.cs

```diff
@@ -749,6 +751,56 @@ public void AcquireTokenSilentServiceErrorTest()
             }
         }
 
+        [TestMethod]
+        public async Task SuggestedCacheKeyTestAsync()
+        {
+            // Arrange
+            using (var httpManager = new MockHttpManager())
+            {
+                PublicClientApplication app = PublicClientApplicationBuilder.Create(TestConstants.ClientId)
+                                                                            .WithAuthority(new Uri(ClientApplicationBase.DefaultAuthority), true)
+                                                                            .WithHttpManager(httpManager)
+                                                                            .WithTelemetry(new TraceTelemetryConfig())
+                                                                            .BuildConcrete();
+
+                var tokenCacheHelper = new TokenCacheHelper();
+                tokenCacheHelper.PopulateCache(app.UserTokenCacheInternal.Accessor);
+                var cacheAccess = app.UserTokenCache.RecordAccess();
+
+                // Act
+                var allAccounts = await app.GetAccountsAsync().ConfigureAwait(false);
+
+                // Assert
+                Assert.IsNull(cacheAccess.LastNotificationArgs.SuggestedCacheKey, "Cannot suggest a key for GetAccounts");
+                Assert.AreEqual(1, allAccounts.Count());
+                cacheAccess.AssertAccessCounts(1, 0);
+
+                // Act
+                string homeAccId = allAccounts.Single().HomeAccountId.Identifier;
+                IAccount accountById = await app.GetAccountAsync(homeAccId).ConfigureAwait(false);
+                cacheAccess.AssertAccessCounts(2, 0);
+
+                // Assert
+                Assert.AreEqual(homeAccId, cacheAccess.LastNotificationArgs.SuggestedCacheKey);
+                Assert.AreEqual(homeAccId, accountById.HomeAccountId.Identifier);
+
+                // Act
+                await app.AcquireTokenSilent(TestConstants.s_scope, accountById).ExecuteAsync().ConfigureAwait(false);
+
+                // Assert
+                cacheAccess.AssertAccessCounts(3, 0);
+                Assert.AreEqual(homeAccId, cacheAccess.LastNotificationArgs.SuggestedCacheKey);
+
+                // Act
+                await app.AcquireTokenSilent(TestConstants.s_scope, accountById.Username).ExecuteAsync().ConfigureAwait(false);
```

Comment:
@jmprieur @jennyf19 - please see this test case - ATS with login hint is not able to suggest a token cache key, because MSAL needs to load all accounts and filter by username. Loading all accounts triggers a cache read. 

If this is not acceptable, we should discuss.

Reply:
Yes, @bgavrilMS : we need to discuss.
this API, as well as GetAccountsAsync() don't make sense for confidential client applications.
but this is a different problem

Reply:
We can duplicate `GetAccountsAsync` methods so that they exist on both PCA and CCA. And then Obsolete them on CCA. But why would I not want to do `GetAccountsAsync` on a cca?